drivers/dai/intel/dmic/dmic.c

```diff
@@ -31,6 +31,82 @@ struct dai_dmic_global_shared dai_dmic_global;
 
 int dai_dmic_set_config_nhlt(struct dai_intel_dmic *dmic, const void *spec_config);
 
+/* Exponent function for small values of x. This function calculates
+ * fairly accurately exponent for x in range -2.0 .. +2.0. The iteration
+ * uses first 11 terms of Taylor series approximation for exponent
+ * function. With the current scaling the numerator just remains under
+ * 64 bits with the 11 terms.
+ *
+ * See https://en.wikipedia.org/wiki/Exponential_function#Computation
+ *
+ * The input is Q3.29
+ * The output is Q9.23
+ */
+static int32_t exp_small_fixed(int32_t x)
```

Comment:
I know this isn't really related to this PR and you're just moving this around, but this code is weirding me out.  The DMIC driver should be copying raw data from the device to an application layer.  If there is rescaling/transformation/resampling/whatever needed that really should be done at the app layer.  Zephyr doesn't (and shouldn't!) know anything about how PDM microphone signal processing works, that's what SOF is for.

Reply:
I think the original reason to have it very close to hardware is that there's a "feature" in dmic hw, that makes a "pop" (unwanted spike in the waveform) in the very beginning. The dmic does some processing in hw ip, like decimation filtering etc. but there's no hw implementation for ramp or any other algorithm to get rid of the error.

So we could have a "ramp component" in sof pipeline, but what it would essentially do is to make a couple of millisecond ramp, then just pass through the 99,9% of the samples.

In this PR I was just moving this away from the header as it was giving me warning from dmic.nhlt.c